blockwork/build/caching.py

```diff
@@ -0,0 +1,159 @@
+    - Use input interface names along with the hashkeys of transforms that 
+        output them to get the transform hashkeys for nodes with dependencies
+
+  Pre-run:
+    - Go through the transforms in reverse order
+    - Try and pull all output interfaces from caches - if successful mark that
+        transform as fetched.
+    - If all dependents of a transform are fetched, mark a transform as skipped
+
+  During-run:
+    - Go through transforms in dependency order as usual
+    - Skip transforms marked as fetchrd or skipped
+    - Push output interfaces to all caches that allow it
+
+Future improvements:
+  - Pass through information such as tags, file-size, and time-to-create 
+    through to caches so they can make intelligent decisions on what to cache.
+
+'''
+
+from abc import ABC, abstractmethod
+import hashlib
+import os
+from pathlib import Path
+from typing import Optional, TYPE_CHECKING
+if TYPE_CHECKING:
+    from blockwork.build.transform import Transform
+from ..context import Context
+
+# Switch to never pull or push from the cache, but to instead compare the 
+# hash with existing cache entries.
+# This will become a command option later...
+CACHE_CONSISTENCY_MODE = False
+
+class Cache(ABC):
+    @staticmethod
+    def hash_content(path: Path) -> str:
+        '''
+        Hash the content of a file or directory. This needs to be consistent
+        across caching schemes so consistency checks can be performed.
+        '''
+        if not path.exists():
+            raise RuntimeError
+        if path.is_dir():
+            content_hash = hashlib.md5('<dir>'.encode('utf8'))
+            for item in sorted(os.listdir(path)):
+                content_hash.update((item + Cache.hash_content(path / item)).encode('utf8'))
+        else:
+            with path.open('rb') as f:
+                content_hash = hashlib.file_digest(f, 'md5')
+        return content_hash.hexdigest()
+
+    @staticmethod
+    def store_all(ctx: Context, key_hash: str, frm: Path) -> bool:
+        'Store to every cache that will accept the content'
+        stored_somewhere = False
+        for cache in ctx.caches:
+            if cache.store(key_hash, frm):
+                stored_somewhere = True
+        return stored_somewhere
+
+    @staticmethod
+    def fetch_any(ctx: Context, key_hash: str, to: Path) -> bool:
+        'Pull from the first cache that has the item'
+        for cache in ctx.caches:
+            if cache.fetch(key_hash, to):
+                return True
+            to.unlink(missing_ok=True)
+        return False
+    
+    @staticmethod
+    def fetch_transform(ctx: Context, transform: "Transform") -> bool:
+        'Fetch all the output interfaces for a transform'
+        if CACHE_CONSISTENCY_MODE:
+            return False
+        
+        for idx, iface in enumerate(transform._flat_output_interfaces):
+            hashkey = f"{iface.get_hashsource(ctx)}-{idx}"
```

Comment:
Are we 100% sure that ordering of output interfaces is reliable enough for the index to be reproducible? (i.e. there's no sets involved, or other non-deterministic collection)

Reply:
Nope, it's something we will need to be-careful/think-more about since even if in bw code the order is reliable user code for meta-interfaces could still be unstable and change it.

(I do however think it's good enough for now)